Workers respond immediately while loading data #381 #447

abyrd opened the pull request:
As explained in #381, workers would receive an HTTP request, then block while loading and preparing all the data needed to perform an analysis. This was done on the initial assumption that data preparation would be fast. In practice it can take up to half an hour or more on large regions like NL or Switzerland.

This PR changes the system so that workers always respond immediately, giving the backend and UI an explanation of why they are not ready if they can't immediately perform the requested analysis.

It will require a corresponding change in the backend that ensures that the HTTP headers and status codes of workers' single point responses are preserved when sent back to the client UI.

Some code has been cleaned up and refactored along the way to be more readable / maintainable.

In this PR's current state, workers often just return text/plain with an error code. This should be cleaned up with a table of expected error codes and JSON response formats.

Methods that will need to be changed / refactored for proper error handling are:
com.conveyal.r5.analyst.cluster.AnalystWorker#reportTaskErrors
com.conveyal.r5.analyst.cluster.AnalystWorker#addErrorJson

ansoncfit commented on src/main/java/com/conveyal/r5/analyst/cluster/TravelTimeSurfaceTask.java:
```diff
@@ -15,6 +14,9 @@
 public class TravelTimeSurfaceTask extends AnalysisTask {
+
+    // FIXME red flag - what is this enum enumerating Java types?
+
```
Should we rename AnalysisTask.Type?

abyrd replied:
It's not that we should rename the enum - it's that Java already has a type system, which allows checking if an object is an `instanceof` a class (which is also a bit of a hack) or even better, allows polymorphism. Different subclasses or implementations should just have methods with the same names that behave differently.

This is probably a remnant of using a single task-enqueue endpoint for both regional and single point tasks. I have created a separate issue for this:  #462

ansoncfit commented:
Are these changes also relevant for #164, #217?

abyrd commented:
> The required "corresponding change in the backend" is to ensure proxies of workers' responses to single point requests keep their headers and status codes, right?

Yes, I updated the PR description to clarify this.

> Are these changes also relevant for #164, #217?

I have read and updated those issues in light of recent commits. I think they'll require some additional work outside this PR.

abyrd commented on src/main/java/com/conveyal/r5/analyst/DataPreloader.java:
```diff
@@ -0,0 +1,239 @@
+             *
+             * Reuse linkages in the base gridPointSet stored in the TransportNetwork as to avoid relinking
+             *
+             * Regional analyses use the extents of the destination opportunity grids as their destination extents.
```
Yes, we should probably add more distinct fields for origin extents and destination extents. This could be aided by the new com.conveyal.r5.analyst.WebMercatorExtents class.

abyrd commented on src/main/java/com/conveyal/r5/analyst/DataPreloader.java:
```diff
@@ -0,0 +1,239 @@
+ * 3. We don't really want to trigger actions after the asynchronous loading is finished, we just want to know
+ *    immediately at the time of each call whether it's finished building or not.
+ *
+ * The steps that need to happen are the following:
```
Javadoc cleaned up in 6146f8db578f1475aecaf5e6f50a16aa7b2da55c

abyrd commented on src/main/java/com/conveyal/r5/analyst/DataPreloader.java:
```diff
@@ -0,0 +1,239 @@
+            // This could easily be done with pointSets.add(network.gridPointSet);
+            // However we might not always want to search out to such a huge destination grid.
+
+            // This is a regional analysis that is not a static site.
```
As I obsoleted code, I was just accumulating useful comments here. Some of them are now unmoored from any relevant code. I have removed them in the comments cleanup in 6146f8db578f1475aecaf5e6f50a16aa7b2da55c

abyrd commented on src/main/java/com/conveyal/r5/analyst/WebMercatorGridPointSetCache.java:
```diff
@@ -9,6 +9,12 @@
+ *
+ * The WebMercatorGridPointSets are very small and fetching one doesn't include linking.
+ * We cache these objects because once they are linked,they contain the linkages, and creating the linkages takes a
+ * lot of time.
```
Comment added in eec071e44d4acf3451470b6a9e84f32cae960546

abyrd commented on src/main/java/com/conveyal/r5/analyst/cluster/AnalystWorker.java:
```diff
@@ -534,7 +569,7 @@ public static void addErrorJson (OutputStream outputStream, List<TaskError> scen
      * @return a list of work tasks, or null if there was no work to do, or if no work could be fetched.
      */
-    public List<AnalysisTask> getSomeWork () {
+    public List<RegionalTask> getSomeWork () {
```
Actually for regional tasks, a blocking call is made to the preloader, so the worker threads just wait there until the network and linkages are ready. The difference is that I simply removed the lines that build the network at all at startup. The worker polls the backend and says what kind of tasks it's meant to handle, and the network is then built when it receives those tasks. By that time it's announced its existence to the backend by polling.

abyrd commented on src/main/java/com/conveyal/r5/analyst/DataPreloader.java:
```diff
@@ -0,0 +1,239 @@
+ *
+ * Created by abyrd on 2018-09-17
+ */
+public class DataPreloader extends AsyncLoader<DataPreloader.Key, TransportNetwork> {
```
Good point. I renamed it to `NetworkPreloader`, in 48390b152d424460e3cb4df2fdf7770feed0afab because it subclasses `AsyncLoader` (loader here referencing the idea of a LoadingCache) and I find it useful to remember that its purpose is to load in advance of any analysis happening (pre-).

This is still not ideal because a lot of the time is not spent on the network (but rather on the linkages and tables). But it's already a bit less generic.

abyrd commented:
Here is a summary of responses you should receive from the backend or worker (via the backend) when issuing a single point request:

| situation                           | response  originates at | HTTP status code | content-type             | body                                                                    |
|-------------------------------------|-------------------------|------------------|--------------------------|-------------------------------------------------------------------------|
| task processed (GRID)               | worker                  | 200              | `application/octet-stream` | Analysis binary grid, followed by scenario application warnings as JSON |
| task processed (GeoTIFF)            | worker                  | 200              | `application/x-geotiff`    | GeoTIFF file (no warnings included)                                     |
| no worker running                   | broker                  | 202              | `application/json`         | `{message: "Starting workers, try again later."}`                         |
| TransportNetwork not ready          | worker                  | 202              | `application/json`         | `{message: "Building network (X% complete)"}`                             |
| error applying scenario             | worker                  | 400              | `application/json`         | `{scenarioApplicationWarnings: [...]}`                                   |
| timeout or error proxying to worker | broker                  | 400              | `application/json`         | `{type:"UNKNOWN", message:..., stackTrace:..}`                            |

I will push changes soon so responses match this table. 

Some thoughts on what I see here: the UI is using `Accept:` headers to select geotiff or grid format. The backend then converts this to a format field within the task object sent to R5. We should be able to preserve this request header all the way back to the worker (instead of using a field on the task), but I'm not sure there's much benefit other than consistency.

When the TransportNetwork is not yet ready, we return a somewhat structured message (containing a completion percentage) but represented as text. I could surface these details but it will require some refactoring, and we will probably eventually have a more sophisticated background tracking system, so I can't yet guess what the final format will be. We may as well just show strings for now and let the worker format them.

I find it a little strange to return 400 codes when there really is a problem with the backend server (e.g. timeout or IO exception while proxying a request to a worker, but I continue to return 400s because all existing code does this.

abyrd commented:
This is open for review - it should be ready for evaluation on staging, along with the accompanying change in conveyal/analysis-backend#183.

abyrd commented on src/main/java/com/conveyal/r5/analyst/NetworkPreloader.java:
```diff
@@ -0,0 +1,241 @@
+ * 3. We don't really want to trigger callbacks after the asynchronous loading is finished, we just want to know
+ *    immediately at the time of each HTTP request whether it's finished building or not.
+ *
+ * Repeatedly re-sending a full Analysis task over HTTP just to see if the network is ready (and failing if it's not)
```
Issuing a separate request to prep the network before firing off the actual analysis request seems in line with our long-term goal of explicitly triggering slow operations.

However in this case, the information needed to prep the network (build / load / link / build distance tables) is several fields in the request itself (network and scenario ID, modes of travel, destination grid dimensions). We'd probably end up sending the same request to the prep-network endpoint as to the analysis endpoint itself.

Consider that once the network, linkages, and distance tables are prepared, they may be evicted from cache on the worker. I'm not saying that's ideal, that's just how it works now. So once you've hit the prep-network endpoint and it says "ready", you may hit the analysis endpoint and learn that it's not ready. Doing it with a single endpoint, once the data are ready they are used immediately for analysis. Having separate endpoints means that you always need to hit two endpoints - first to prep the network, then to analyze - for every single point request, and you may need to repeat the cycle if the data are no longer available when you hit the analysis endpoint. It would generate more traffic and not provide any immediate advantage.

Longer term what we want is for the single point analysis endpoint to do only one thing: deserialize the network, linkage, and distance tables from disk. That would be quite fast, on the order of a few seconds. Those serialized, saved data tables would be generated in a separate step. 

So as a development strategy, I'd like to keep the current system (send the task directly to the single-point endpoint without first hitting another endpoint to prime the data), but completely factor out all the graph build, linkage, and distance table calculations into a manually triggered build step that would have its own endpoint. That would have to be done later though.

abyrd commented:
@ansoncfit I have removed the percentage progress information from the async loader responses, and I resolved merge conflicts with dev. Deploying this PR to staging will now also include serialization changes.

abyrd commented:
@ansoncfit the merge was a little more complicated than expected. One last thing for your approval, see commit 5642460 I just added.

ansoncfit commented:
Thanks for catching and fixing the fare calculator issue @abyrd.  I'll merge now.

ansoncfit commented:
:tada: This PR is included in version 4.3.0 :tada:

The release is available on [GitHub release](https://github.com/conveyal/r5/releases/tag/v4.3.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
